homeassistant/components/velbus/diagnostics.py

```diff
@@ -0,0 +1,57 @@
+"""Diagnostics support for Velbus."""
+from __future__ import annotations
+
+from typing import Any
+
+from velbusaio.module import Module as VelbusModule
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.device_registry import DeviceEntry
+
+from .const import DOMAIN
+
+
+async def async_get_config_entry_diagnostics(
+    hass: HomeAssistant, entry: ConfigEntry
+) -> dict[str, Any]:
+    """Return diagnostics for a config entry."""
+    controller = hass.data[DOMAIN][entry.entry_id]["cntrl"]
+    data: dict[str, Any] = {"entry": entry.as_dict(), "modules": []}
+    for module in controller.get_modules().values():
+        data["modules"].append(build_module_diagnostics_info(module))
+    return data
+
+
+async def async_get_device_diagnostics(
+    hass: HomeAssistant, entry: ConfigEntry, device: DeviceEntry
+) -> dict[str, Any]:
+    """Return diagnostics for a device entry."""
+    controller = hass.data[DOMAIN][entry.entry_id]["cntrl"]
+    channel = list(next(iter(device.identifiers)))[1]
+    modules = controller.get_modules()
+    return build_module_diagnostics_info(modules[int(channel)])
+
+
+def build_module_diagnostics_info(module: VelbusModule) -> dict[str, Any]:
+    """Build per module diagnostics info."""
+    data: dict[str, Any] = {
+        "type": module.get_type_name(),
+        "address": module.get_addresses(),
+        "name": module.get_name(),
+        "sw_version": module.get_sw_version(),
+        "is_loaded": module.is_loaded(),
+        "channels": build_channels_diagnostics_info(module.get_channels()),
+    }
+    return data
+
+
+def build_channels_diagnostics_info(channels: dict[str, Any]) -> dict[str, Any]:
+    """Build diagnostics info for all channels."""
+    data: dict[str, Any] = {}
+    for channel in channels.values():
+        data[channel.get_channel_number()] = {}
+        for key, value in channel.__dict__.items():
```

Comment:
Why not access the channel attributes that we're interested in directly? We can create a set or a dict with the interesting attribute names (and names to store them under) and use `getattr` to get them. I don't think we should look in `__dict__`.

Reply:
the attributes available depend on the channel types, depending of the channeltype we can have a list of attributes available.
With doing it this way there is no need to maintain this list on 2 places, once in the velbusaio module and once in hass diagnostics

Reply:
Then I suggest you create a method on the channel to explicitly export the interesting info. The library consumer shouldn't dig into instance details like this.